06_IPC/03_Message_Passing.md

```diff
@@ -0,0 +1,176 @@
+
+All that's left is to tell the receiver it has a message is available by placing the buffer address on the endpoint. Again, notice the use of the lock to prevent race conditions while we mess with the internals of the endpoint.
+
+```c
+acquire(&target->lock);
+target->msg_buffer = msg_copy;
+target->msg_length = length;
+release(&target->lock);
+```
+
+After the lock on the endpoint is released, the sent has been sent! Now it's up to the receiving thread to check the endpoint and set the buffer to NULL again.
+
+### Multiple Messages
+
+In theory this works, but we've overlooked one huge issue: what if there's already a message at the endpoint? You should handle this, and there's a couple of ways to go about it:
+
+- Allow for multiple messages to be stored on an endpoint.
+- Fail to send the message, instead returning an error code from `ipc_send()`.
+
+The first option is recommended, as it's likely there will be some processes that handle a lot of messages. Implementing this is left as an exercise to the user, but a simple implemenation might use a struct to hold each message (the buffer address and length) and a next field. Yes, more linked lists!
+
+Sending messages would now mean appending to the list instead of writing the buffer address as before. 
+
+## Receiving
+
+We can send messages, next let's look at receiving them. We're going to use a basic (and inefficient) example, but it shows how it could be done. 
+
+The theory behind this is simple: when we're in the receiving process, we allocate a buffer to hold the message, and copy the messge data stored at the endpoint into our local buffer. Now we can set the endpoint's `msg_buffer` field to `NULL` to indicate that there is no longer a message to be received. Note that setting the buffer to `NULL` is specific to our example code, and your implementation may be different.
+
+As always, note the use of locks to prevent race conditions. The variable `endpoint` is assumed to be the endpoint we want to receive from.
+
+```c
+ipc_endpoint* endpoint;
+
+acquire(&endpoint->lock);
+void* local_copy = malloc(endpoint->msg_length);
+memcpy(local_copy, endpoint->msg_data, endpoint->msg_length);
+
+endpoint->msg_data = NULL;
+endpoint->msg_length = 0;
+release(&endpoint->lock)'
+```
+
+At this point the endpoint is now ready to receive another message, and we've got a copy of the message in `local_copy`. You're successfully passed a message from one address space to another!
+
+## Additional Notes
+
+- We've described a double-copy implementation here, but you might want to try a single-copy implemenation. Single-copy implementations *can* be faster, but they require extra logic. For example the kernel will need to access the recipient's address space from the sender's address space, how do you manage this? If you have all of physical memory mapped somewhere (like an identity map) you could use this, otherwise you will need some way to access this memory.
```

Comment:
not sure what's going on here, but the sintax from the github syntax viewer is strange, try to double check if there are some markdown issues.

Reply:
It was a misplaced single quote (') instead of a semicolon on the previous code block :)